core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryBuilder.java

```diff
@@ -394,4 +395,17 @@ public FilterFunctionBuilder readFrom(StreamInput in) throws IOException {
             return new FilterFunctionBuilder(in.readQuery(), in.readScoreFunction());
         }
     }
+
+    @Override
+    public QueryBuilder<?> rewrite(QueryRewriteContext queryShardContext) throws IOException {
+        QueryBuilder<?> queryBuilder = this.query.rewrite(queryShardContext);
+        if (queryBuilder != query) {
```

Comment:
good point! I think we need to iterate over the filterFunctionBuilders and rewrite their corresponding filters